client/src/app/hooks/table-controls/useLocalTableControlState.ts

```diff
@@ -77,3 +87,75 @@ export const useLocalTableControlState = <
     currentPageItems: hasPagination ? currentPageItems : sortedItems,
   };
 };
+
+// TODO refactor useUrlParams so it can be used conditionally (e.g. useStateOrUrlParams) so we don't have to duplicate all this.
+//      this would mean all use[Feature]UrlParams hooks could be consolidated into use[Feature]State with a boolean option for whether to use URL params.
+
+export const useLocalTableControlUrlParams = <
+  TItem,
+  TColumnKey extends string,
+  TSortableColumnKey extends TColumnKey,
+  TURLParamKeyPrefix extends string = string,
+>(
+  args: IUseLocalTableControlStateArgs<TItem, TColumnKey, TSortableColumnKey> &
+    IExtraArgsForURLParamHooks<TURLParamKeyPrefix>
+): IUseTableControlPropsArgs<TItem, TColumnKey, TSortableColumnKey> => {
+  const {
+    items,
+    filterCategories = [],
+    sortableColumns = [],
+    getSortValues,
+    initialSort = null,
+    hasPagination = true,
+    initialItemsPerPage = 10,
+    idProperty,
+    initialSelected,
+    isItemSelectable,
+  } = args;
+
+  const filterState = useFilterUrlParams(args);
+  const { filteredItems } = getLocalFilterDerivedState({
+    items,
+    filterCategories,
+    filterState,
+  });
+
+  const selectionState = useSelectionState({
+    items: filteredItems,
+    isEqual: (a, b) => a[idProperty] === b[idProperty],
+    initialSelected,
+    isItemSelectable,
+  });
+
+  const sortState = useSortUrlParams({ ...args, sortableColumns, initialSort });
```

Comment:
You _could_ use a bit of trickery to get this as a single call .....

```jsx
  const isUrlParamBased = !!urlParamKeyPrefix;

  const sortState = (isUrlParamBased ? useSortUrlParams : useSortState)({
    ...args, 
    sortableColumns,
    initialSort,
  });
```

Reply:
I considered this, but the eslint rules-of-hooks rule doesn't like it I think (if you re-render with a different value for `isUrlParamBased` it would switch up which hooks are called and screw up the render, not that we'd do that). If we added an eslint-ignore there I think it would be too easy to accidentally break it in a future change.

What I'd like to do eventually (soon perhaps) is refactor `useUrlParams` (which is used inside `useSortUrlParams` and each of the other `use[Feature]UrlParams` hooks) so that it itself has a `isUrlParamBased` condition, and has its own `useState` internally that it either uses or ignores in favor of URL params depending on config. Then we could get rid of all the `use[Feature]UrlParams` hooks and instead have the URL params be an option in `use[Feature]State`, all the way up to `useTableControlState`.